src/style-spec/function/expressions.js

```diff
@@ -0,0 +1,354 @@
+        compile: () => ({
+            js: '(feature.geometry || {}).type || null',
+            isFeatureConstant: false
+        })
+    },
+    'id': {
+        name: 'id',
+        type: lambda(ValueType),
+        // TODO: should yield error if missing
+        compile: () => ({
+            js: 'typeof feature.id === "undefined" ? null : feature.id',
+            isFeatureConstant: false
+        })
+    },
+    'zoom': {
+        name: 'zoom',
+        type: lambda(NumberType),
+        compile: () => ({js: 'mapProperties.zoom', isZoomConstant: false})
+    },
+    '+': defineBinaryMathOp('+'),
+    '*': defineBinaryMathOp('*'),
+    '-': defineBinaryMathOp('-'),
+    '/': defineBinaryMathOp('/'),
+    '%': defineBinaryMathOp('%'),
+    '^': {
+        name: '^',
+        type: lambda(NumberType, NumberType, NumberType),
+        compile: (_, args) => ({js: `Math.pow(${args[0].js}, ${args[1].js})`})
+    },
+    'log10': defineMathFunction('log10', 1),
+    'ln': defineMathFunction('ln', 1, 'log'),
+    'log2': defineMathFunction('log2', 1),
+    'sin': defineMathFunction('sin', 1),
+    'cos': defineMathFunction('cos', 1),
+    'tan': defineMathFunction('tan', 1),
+    'asin': defineMathFunction('asin', 1),
+    'acos': defineMathFunction('acos', 1),
+    'atan': defineMathFunction('atan', 1),
+    '==': defineComparisonOp('=='),
+    '!=': defineComparisonOp('!='),
+    '>': defineComparisonOp('>'),
+    '<': defineComparisonOp('<'),
+    '>=': defineComparisonOp('>='),
+    '<=': defineComparisonOp('<='),
+    '&&': defineBooleanOp('&&'),
+    '||': defineBooleanOp('||'),
+    '!': {
+        name: '!',
+        type: lambda(BooleanType, BooleanType),
+        compile: (_, args) => ({js: `!(${args[0].js})`})
+    },
+    'upcase': {
+        name: 'upcase',
+        type: lambda(StringType, StringType),
+        compile: (_, args) => ({js: `(${args[0].js}).toUpperCase()`})
+    },
+    'downcase': {
+        name: 'downcase',
+        type: lambda(StringType, StringType),
+        compile: (_, args) => ({js: `(${args[0].js}).toLowerCase()`})
+    },
+    'concat': {
+        name: 'concat',
+        type: lambda(StringType, nargs(ValueType)),
+        compile: (_, args) => ({js: `[${args.map(a => a.js).join(', ')}].join('')`})
+    },
+    'rgb': {
+        name: 'rgb',
+        type: lambda(ColorType, NumberType, NumberType, NumberType),
+        compile: fromContext('rgba')
+    },
+    'rgba': {
+        name: 'rgb',
+        type: lambda(ColorType, NumberType, NumberType, NumberType, NumberType),
+        compile: fromContext('rgba')
+    },
+    'case': {
+        name: 'case',
+        type: lambda(typename('T'), nargs(BooleanType, typename('T')), typename('T')),
+        compile: (_, args) => {
+            args = [].concat(args);
+            const result = [];
+            while (args.length > 1) {
+                const c = args.splice(0, 2);
+                result.push(`${c[0].js} ? ${c[1].js}`);
```

Comment:
👍 yep -- it's done globally at the end of `compile()`, https://github.com/mapbox/mapbox-gl-js/pull/4841/commits/04dbb977691d2a1710e0587c6e8cb058138ebf12#diff-442324b3174c25e635f18536d7732df2R147, but this reminds me, we should include a test that targets this since it's be very easy to regress